### PR TITLE
Clearer error message if async context manager used synchronously

### PR DESCRIPTION
**Pull request by Carreau**
Otherwise it says it does not have `__enter__` which is obvious for
advanced pythonista. Still a tiny bit clearer is likely better.

I guess this _might_ fool static type checkers, but I'm unaware of any
that would flag that.

### CODE REVIEW

File: trio/_util.py [hunk 1]
```diff
@@ -135,6 +136,13 @@ def __init__(self, func, args, kwds):
                 if sys.exc_info()[1] is not value:
                     raise
 
+    def __enter__(self):
+        raise RuntimeError("The function `{}` create an async context manager, use with `async with` and not `with`.".format(self._func_name))
```
Comment (njsmith) [inline trio/_util.py]:
```python
"use 'async with {func_name}(...)', not 'with {func_name}(...)'".format(name=self._func_name)"
```
(and 80 column limit!)

File: trio/_util.py [hunk 2]
```diff
@@ -135,6 +136,13 @@ def __init__(self, func, args, kwds):
...
+
+    def __exit__(self):
```
Comment (njsmith) [inline trio/_util.py]: `# pragma: no-cover`

**Comment by Carreau** [comment]
Hum, also unawaited coroutines. I need to fix the test.

**Comment by njsmith** [comment]
If it does create a problem for static type checkers then it should be fixable by adding an explicit annotation to `acontextmanager`.

Something else we might want to think about is coming up with a more standard/comprehensive way of doing this kind of checking, like a decorator we can put on all our context managers and iterators. But that doesn't need to block this; this case is a bit special because we can give a nice error message.

File: trio/_util.py [hunk 3]
```diff
@@ -135,6 +136,13 @@ def __init__(self, func, args, kwds):
                 if sys.exc_info()[1] is not value:
                     raise
 
+    def __enter__(self):
+        raise RuntimeError("use 'async with {func_name}(...)', not 'with {func_name}(...)'".format(func_name=self._func_name))
+
+    def __exit__(self):
+        assert False, """Never called, but should be defined""" # pragma: no-cover
```
Comment (njsmith) [inline trio/_util.py]: Should be `no cover`, not `no-cover`; it's not actually being recognized right now :-). Also better to put two spaces before the `#`, and to move it up one line, so:
```python
def __exit__(self):  # pragma: no cover
```

File: trio/_core/tests/test_run.py [hunk 4]
```diff
@@ -89,6 +89,16 @@ def test_run_nesting():
         await task.wait()
         assert task.result.unwrap() == 20
 
+async def test_nursery_warn_use_async_with():
+    with pytest.raises(RuntimeError) as excinfo:
+        on = _core.open_nursery()
+        with on as nursery:
+            pass # pragma: no-cover
```
Comment (njsmith) [inline trio/_core/tests/test_run.py]: Also needs an extra space before `#` and removal of the `-`

**Comment by njsmith** [comment]
Oo, I figured out how to fix the nitpicks through github. Let's see what the CI thinks...

**Comment by njsmith** [comment]
Thanks!

**Comment by Carreau** [comment]
Thanks !

**Pull request merged**